airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -71,6 +75,66 @@ def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
         return self._state
 
 
+class CursorBasedCheckpointReader(CheckpointReader):
+    """
+    CursorBasedCheckpointReader is used by streams that implement a Cursor in order to manage state. This allows the checkpoint
+    reader to delegate the complexity of fetching state to the cursor and focus on the iteration over a stream's partitions.
+    Right now only low-code connectors provide cursor implementations, but the logic is extensible to any stream that adheres
+    to the Cursor interface.
+    """
+
+    def __init__(self, cursor: Cursor, stream_slices: Iterable[Optional[Mapping[str, Any]]], read_state_from_cursor: bool = False):
+        # The first attempt of an RFR stream has an empty {} incoming state, but should still make a first attempt to read records
+        # from the first page in next().
+        self._cursor = cursor
+        self._stream_slices = iter(stream_slices)
+        self._read_state_from_cursor = read_state_from_cursor
+        self._current_slice: Optional[StreamSlice] = None
+        self._finished_sync = False
+
+    def next(self) -> Optional[Mapping[str, Any]]:
```

Comment:
This method is a bit more complicated than it needs to be for streams because it can iterate in two dimensions: over incoming static set of slices and dynamically based on the current cursor's stream state.

Since this is only scoped to full refresh streams (not substreams  nor incremental), there should only be one static slice to loop over). But we support it regardless and positions us better for substream low-code RFR when we prioritize it.